File: Java Programs/SortingAlgos/MergeSort.java
import java.util.*;

class MergeSort {

    public static int[] mergeSort(int[] arr,int start,int end) {
        if(start < end) {
            int mid = (start + end) / 2;
            mergeSort(arr,start,mid);
            mergeSort(arr,mid+1, end);
            merge(arr,start, mid,end);
        }

        return arr;
    }

    public static void merge(int[] arr, int start, int mid, int end) {
        int[] temp = new int[end - start + 1];
        int i = start;
        int j = mid + 1;
        int k = 0;

        while((i <= mid) && (j <= end)) {
            if(arr[i] <= arr[j]) {
                temp[k] = arr[i];
                k++;
                i++;
            }
            else {
                temp[k] = arr[j];
                k++;
                j++;
            }
        }

        while(i <= mid) {
            temp[k] = arr[i];
            k++;
            i++;
        }

        while(j <= end) {
            temp[k] = arr[j];
            k++;
            j++;
        }

        for(int a = start; a< end+1; a++) {
            arr[a] = temp[a - start];
        }
    }

    public static void main(String[] args) {
        int[] arr = {7,8,3,1,2,-1,-4,0};

        mergeSort(arr,0,arr.length -1);

        for(int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }
}

File: Python/Recursion/MergeSort.py
def MergeSort(arr, start, end):
    if(start < end):
        mid = (start + end) // 2
        MergeSort(arr, start, mid)
        MergeSort(arr, mid + 1, end)
        merge(arr , start, mid, end)

    return arr

def merge(arr, start,mid, end):
    temp = [0]*(end - start + 1) # size of this array woulf be [end - start + 1]
    i = start
    j = mid + 1
    k = 0

    while((i<=mid) and (j <= end)):
        if arr[i] <= arr[j]:
            temp[k] = arr[i]
            k += 1
            i += 1
        else:
            temp[k] = arr[j]
            k += 1
            j += 1

    while(i <= mid):
        temp[k] = arr[i]
        k += 1
        i += 1

    while(j <= end):
        temp[k] = arr[j]
        k += 1
        j += 1

    for i in range(start, end + 1):
        arr[i] = temp[i - start]

if __name__ == '__main__':
    arr = [-5,20,10,3,2,0]
    print(MergeSort(arr, 0, len(arr)-1))